### ISSUE TITLE
`bag.map(str)` fails in Python 3.4

### ISSUE THREAD

**Issue by StuartAxelOwen** [ContinuumIO/dask#472]
`' '.join(bag.from_sequence(range(4)).map(str))` should yield `0 1 2 3`, but instead fails in function application.  This seems to be caused by `dask.bag.core.takes_multiple_arguments`, which asserts that `str` takes multiple arguments (which it can), causing `dask.compatibility.apply` to attempt to use *args, when args is potentially not a list.

This isn't a problem in Python 2, as `inspect.getargspec` raises a TypeError for classes.

**Comment by StuartAxelOwen** [comment]
This is also a problem for classes, and particularly classes inheriting from `object`, as the argspec for their __init__ function is `ArgSpec(args=['self'], varargs='args', keywords='kwargs', defaults=None)`

**Comment by mrocklin** [comment]
Yes, that seems like a genuine issue and a good synopsis.  I can take a look at this early tomorrow.

**Comment by mrocklin** [comment]
(That is if, of course, someone doesn't beat me to it :))

**Comment by StuartAxelOwen** [comment]
Pull request doesn't completely solve the problem - I'm not sure how we can correctly get the argspec for `object`-based classes, but fixes it for other cases.

**Comment by StuartAxelOwen** [comment]
We could potentially just check function membership in a set of builtins known to be single arity, which does better than the previous attempt I took - but that still doesn't solve the "inherits from object" problem.

**Comment by StuartAxelOwen** [comment]
Okay, was probably drunk - using `inspect.getargspec` works for classes inheriting from `object`, I'll have update the PR and reopen it.

**Comment by jcrist** [comment]
Another solution would be to remove the magic in `map`, in favor of `map` and `starmap` (as itertools does it). This removes any need for introspection, is explicit, and is already a proven interface.

**Comment by StuartAxelOwen** [comment]
I like that suggestion - introspection here creates a lot of corner cases where `map` can get it wrong and the user would have no recourse. Also, the zen of Python seems pretty specific here:

> In the face of ambiguity, refuse the temptation to guess.

**Comment by mrocklin** [comment]
I would be comfortable removing the guesswork if people think that that's best.  The one concern I have is that Python 3 no longer supports argument unpacking in lambda signatures

```python
In [1]: lambda (k, v): v
  File "<ipython-input-1-8a2b95af522d>", line 1
    lambda (k, v): v
           ^
SyntaxError: invalid syntax
```

In cases where we unambiguously have multiple arguments and all of them must be used (multiple non-variadic args) then I think it might still be useful to maintain this convenience.

**Comment by jcrist** [comment]
I don't see how that can't be done with `starmap`.

`map(f, itbl) -> [f(i) for i in itbl]`
`starmap(f, itbl)` -> [f(*args) for args in itbl]`

If the magic is removed from `map`, starmap simply becomes the following:

```
def starmap(self, f):
    return self.map(lambda args: f(*args))
```

In this way, it's up to the user to know if they want to splat the args in the bag into the function or not.

**Comment by mrocklin** [comment]
I've found that the number of users who know about `map` vastly outweigh the number of users who know about `starmap`.  

My suggestion would be that we respect Stuart's zen of Python quote

> In the face of ambiguity, refuse the temptation to guess.

And only use `apply` when it is unambiguously the user's intent as in the case of `lambda a, b: ...`

**Comment by jcrist** [comment]
Saying "many users don't know about it, so we shouldn't implement it"  seems like a bad philosophy to me - good docs and examples would be a solution to that. This also breaks compatibility with python's `map` api, which I think is how people would expect it to work. It's definitely how I expected it to work, I didn't even know we supported magic splatting. 

I really prefer the explicitness of saying "`map` never splats, `starmap` always splats", and avoiding all the introspection complications. However, if people overwhelmingly want the old functionality, I think the only sane thing to do is only use `apply` when there are no varargs.

**Comment by StuartAxelOwen** [comment]
Would it make sense to add an optional `star` arg to `map`, to allow the user to be explicit in what they want?  This would help the "no one knows about starmap" problem, and allow the user to be explicit.

**Comment by mrocklin** [comment]
I've used dask.bag extensively and have found the auto-splatting to be really convenient, especially when in Python 3.  I've tried to push `starmap` in other projects before and my experience is that the `starmap` abstraction is virtually unknown.  While I appreciate that we have the ability to push a bit on people's knowledge of abstractions this ability is *very very slight*, happens slowly, and comes at a significant cost.  I haven't found that documentation and examples are sufficient for this, especially when people's patience to read documentation is finite.

However, Stuart's usage experience was that auto-splatting caused a problem.  I think that we can satisfy his experience by being more restrictive with auto-splatting.  This maintains the use cases that I've run into above.

@jcrist If we were to remove this feature then I would want to see that people that have run into this in a bad way in actual problems.  I hesitate to remove a feature that is known to be useful and is cheap to keep around without significant counter-examples coming from real-world-ish use cases.
